### ISSUE TITLE
about ’runtime(android) wenet.cc‘ code optimization，关于代码优化的一点疑问

### ISSUE THREAD

**Issue by 731942771** [wenet-e2e/wenet#1343]
I modified wenet/runtime/android/wenet/app/src/main/cpp/wenet.cc code, I don't know if there is anything wrong.  The decode results of the 'kEndpoint' tag after a pause in speech are returned after about 2 seconds, the accuracy is not higher or lower.
我修改了wenet.cc中的代码，但不确定是否合适，因为简单测试识别结果精确度并没有更高或更低，说话停顿后’kEndpoint‘标记的识别结果都是约2秒后返回。
```
// 版权归属 Copyright (c) 2021 Mobvoi Inc (authors: Xiaoyu Chen)
// 开源协议 Licensed under the Apache License, Version 2.0 (the "License");
// 开源地址 https://github.com/wenet-e2e/wenet/blob/main/runtime/android/wenet/app/src/main/cpp/wenet.cc

#include <jni.h>

#include "torch/script.h"
#include "torch/torch.h"
#include "decoder/asr_decoder.h"
#include "decoder/torch_asr_model.h"
#include "frontend/feature_pipeline.h"
#include "frontend/wav.h"
#include "post_processor/post_processor.h"
#include "utils/log.h"
#include "utils/string.h"

std::shared_ptr<wenet::DecodeOptions> decode_config;
std::shared_ptr<wenet::FeaturePipelineConfig> feature_config;
std::shared_ptr<wenet::FeaturePipeline> featurePipeline;
std::shared_ptr<wenet::AsrDecoder> asrDecoder;
std::shared_ptr<wenet::DecodeResource> resource;
wenet::DecodeState state = wenet::kEndBatch;
std::string final_result;
static bool isRunning = false;

// 初始化资源 init src
void init(JNIEnv *env, jobject obj, jstring jModelDir) {
    const char *pModelDir = env->GetStringUTFChars(jModelDir, nullptr);

    std::string modelPath = std::string(pModelDir) + "/final.zip";
    std::string dictPath = std::string(pModelDir) + "/units.txt";
    auto model = std::make_shared<wenet::TorchAsrModel>();
    model->Read(modelPath);
    LOG(INFO) << "model path: " << modelPath;

    resource = std::make_shared<wenet::DecodeResource>();
    resource->model = model;
    resource->symbol_table = std::shared_ptr<fst::SymbolTable>(fst::SymbolTable::ReadText(dictPath));
    LOG(INFO) << "dict path: " << dictPath;

    wenet::PostProcessOptions post_process_opts;
    resource->post_processor = std::make_shared<wenet::PostProcessor>(post_process_opts);

    feature_config = std::make_shared<wenet::FeaturePipelineConfig>(80, 16000);
    featurePipeline = std::make_shared<wenet::FeaturePipeline>(*feature_config);

    decode_config = std::make_shared<wenet::DecodeOptions>();
    decode_config->chunk_size = 16;
    asrDecoder = std::make_shared<wenet::AsrDecoder>(featurePipeline, resource, *decode_config);
}

// 重置资源 reset src
void reset(JNIEnv *env, jobject obj) {
    asrDecoder->Reset();
    state = wenet::kEndBatch;
    final_result = "";
    isRunning = false;
}

// 识别线程。这个执行kEndpoint结果比较慢，约2秒
void decode_thread_func() {
    while (isRunning) {
        state = asrDecoder->Decode();
        if (state == wenet::kEndFeats || state == wenet::kEndpoint) {
            asrDecoder->Rescoring();
        }

        // add_1

        if (state == wenet::kEndFeats) { // result fast, but inaccurate
            break;
        } else if (state == wenet::kEndpoint) { // result slowly, about 2s

            // add_2 把识别代码放在这里好像也可以，没提高也没降低识别精度，也没降低循环速度。
            // It seems OK decode here,accuracy is neither improved nor reduced, speed not slower or faster
            std::string result;
            if (asrDecoder->DecodedSomething()) {
                result = asrDecoder->result()[0].sentence;
            }

            LOG(ERROR) << "识别：" << result;
            final_result = result;
            asrDecoder->ResetContinuousDecoding();
        } else {
            //asrDecoder->DecodedSomething(); // 测试这里不执行也影响结果。looks like do nothing is ok.
        }
    }
    LOG(ERROR) << "线程结束";
}

// 开始识别
void start_decode(JNIEnv *env, jobject obj) {
    isRunning = true;
    std::thread decode_thread(decode_thread_func); // 启动识别线程
    decode_thread.detach();                        // 识别线程驻留后台运行
}

// 接收语音pcm数据，识别
void accept_waveform(JNIEnv *env, jobject, jshortArray jWaveform) {
    jsize size = env->GetArrayLength(jWaveform);
    int16_t *waveform = env->GetShortArrayElements(jWaveform, 0);
    featurePipeline->AcceptWaveform(waveform, size);
}

// 提取识别结果
jstring get_result(JNIEnv *env, jobject obj) {
    if (final_result.size() > 0) {
        std::string b = final_result.c_str();
        final_result = "";
        return env->NewStringUTF(b.c_str());
    }
    return env->NewStringUTF("");
}

// 停止识别
void stop_decode(JNIEnv *env, jobject obj) {
    isRunning = false; // shutdown thread
    LOG(ERROR) << "停止解码";
    featurePipeline->set_input_finished();
}

jboolean is_finished(JNIEnv *env, jobject obj) {
    if (state == wenet::kEndFeats) {
        return JNI_TRUE;
    }
    return JNI_FALSE;
}

JNIEXPORT jint JNI_OnLoad(JavaVM *vm, void *) {
    JNIEnv *env;

    if (vm->GetEnv(reinterpret_cast <void **>(&env), JNI_VERSION_1_6) != JNI_OK) {
        return JNI_ERR;
    }

    jclass c = env->FindClass("com/cuiweiyou/wenet/tool/WenetTool");
    if (c == nullptr) {
        return JNI_ERR;
    }

    static const JNINativeMethod methods[] = {
            {"init",           "(Ljava/lang/String;)V", reinterpret_cast<void *>( init)},
            {"reset",          "()V",                   reinterpret_cast<void *>( reset)},
            {"stopDecode",     "()V",                   reinterpret_cast<void *>( stop_decode)},
            {"startDecode",    "()V",                   reinterpret_cast<void *>( start_decode)},
            {"acceptWaveform", "([S)V",                 reinterpret_cast<void *>( accept_waveform)},
            {"getResult",      "()Ljava/lang/String;",  reinterpret_cast<void *>( get_result)},
            {"isFinished",     "()Z",                   reinterpret_cast<void *>( is_finished)},
    };
    int rc = env->RegisterNatives(c, methods, sizeof(methods) / sizeof(JNINativeMethod));

    if (rc != JNI_OK) {
        return rc;
    }

    return JNI_VERSION_1_6;
}
```
Please correct it.
希望看到的朋友能进行指正。

**Comment by 731942771** [comment]
尝试了AISHELL-1和Multi-CN两个模型，汉语 后者的结果更精确些。
I tried AISHELL-1 and Multi-CN, the result of Chinese Multi-CN is more accurate.

**Comment by pengzhendong** [comment]
请通过提 pr 的方式，这样看不出来修改了什么地方。

**Comment by 731942771** [comment]
> 请通过提 pr 的方式，这样看不出来修改了什么地方。
谢谢。不用提交pr了，对代码逻辑未做任何修改，只是略作简化了。
可以的话，希望管理员关闭这个issues。
